docs/architecture/adr-d001-inter-validator-set-messaging.md

```diff
@@ -0,0 +1,243 @@
+
+`EndBlock` message shall be extended by adding a network address of each Validator node. Based on that, each
+member of the active Validator Set will connect with a predefined number of Peer Validators belonging to the same 
+Validator Set. A list of Peer Validators will be determined using the algorithm described in 
+[DIP 0006](https://github.com/dashpay/dips/blob/master/dip-0006.md#building-the-set-of-deterministic-connections).
+
+This solution introduces a risk of delayed propagation of new block due to limited connection slots between Full and
+Validator nodes. To mitigate this risk, we can tune the number of Full node connection slots independently from
+Validator node connection slots.
+
+### Scenario 2: Trigger rotation from ABCI app passing quorum id to rotate
+
+This scenario is very similar to Scenario 1. However, instead of passing addresses of all Validators that are members of
+a Validator Set, only quorum ID is passed.
+
+This solution requires Tendermint to determine Validators addresses, effectively re-implementing logic already done in
+the ABCI App.
+
+### Scenario 3: Rotate on Tenderdash
+
+ABCI application can send the whole master node list and quorum list to Tenderdash and
+potentially rotate on its side. However, it would split the responsibility for the
+management of Validator Set rotation between multiple components. It would also require
+much more additional implementation. It would be better to keep the current Tendermint
+design and concept being a logic-less consensus library and keep all business logic in
+the ABCI application.
+
+## Decision
+
+After discussion, we will implement the solution described in **Scenario 1**. It will allow us to:
+
+* reuse logic already implemented as part of ABCI App,
+* keep Tenderdash a logic-less consensus library.
+
+## Detailed Design
+
+### What are the user requirements?
+
+Having a network with much more Full nodes than Validators ensure each Validator Node can connect (directly or
+through another Validator) to any other Validator which is a member of the same active Validator Set.
+
+### What systems will be affected?
+
+* Tenderdash
+* ABCI App
+
+### What new data structures are needed, what data structures need changes?
+
+1. Separate Validator Set member connection pool from Full node connection pool
+2. Additional tuning parameters:
+    1. Full node connection slots
+    2. Validator node connection slots
+
+### What new APIs will be needed, what APIs will change?
+
+In the ABCI protocol, we add a network address of each member of the active Validator Set to the
+`ResponseEndBlock.ValidatorSetUpdate.ValidatorUpdate` structure. The network address structure is as follows:
+
+```protobuf
+enum NetAddressType {
+  IPv4 = 0;
+};
+
+message NetAddress {
+  reserved 1; // Reserved for compatibility with tendermint.p2p.NetAddress
+  reserved "id"; // Reserved for compatibility with tendermint.p2p.NetAddress
+  string address = 2;
```

Comment:
maybe use a uint128 instead?

Reply:
Probably, it's better to keep as a string for the future support with IPv6?

Reply:
ipv6 uses 128 bits, so that's why we can use uint128.

Reply:
I would just go with bytes. Actually that's what someone else is doing too https://stackoverflow.com/questions/51395679/what-is-the-most-efficient-protobuf-type-in-c-for-storing-ipv4-or-ipv6-addre

Reply:
I agree with Sam. Using bytes looks like the most unified approach.